source/PlayerInfo.cpp

```diff
@@ -1356,7 +1367,17 @@ void PlayerInfo::SetShipOrder(const vector<shared_ptr<Ship>> &newOrder)
 	// Check if the incoming vector contains the same elements
 	if(std::is_permutation(ships.begin(), ships.end(), newOrder.begin()))
 	{
+		shared_ptr<Ship> oldFirstShip = ships.front();
 		ships = newOrder;
+		shared_ptr<Ship> newFirstShip = ships.front();
+		// Check if the position of the flagship has changed and
+		// if the ship in the first position can be a flagship and is in the current system.
+		if(newFirstShip != oldFirstShip && Preferences::Has("Automatically unpark flagship")
+				&& newFirstShip->CanBeFlagship() && newFirstShip->GetSystem() == system)
+		{
+			newFirstShip->SetIsParked(false);
+			oldFirstShip->SetIsParked(true);
+		}
```

Comment:
We don't need to clone the shared_ptr here
```suggestion
		Ship *oldFirstShip = ships.front().get();
		ships = newOrder;
		Ship *newFirstShip = ships.front().get();
		// Check if the position of the flagship has changed, the ship in the first position can be
		// a flagship, and is in the current system.
		if(newFirstShip != oldFirstShip && Preferences::Has("Automatically unpark flagship")
				&& newFirstShip->CanBeFlagship() && newFirstShip->GetSystem() == system)
		{
			newFirstShip->SetIsParked(false);
			oldFirstShip->SetIsParked(true);
		}
```